rfcs/20201105-attribute-num-threads/README.md

```diff
@@ -0,0 +1,272 @@
+    ...
+
+    stream(const engine &aengine, flags aflags, const stream_attr &attr);
+
+    ...
+};
+} // namespace dnnl
+~~~
+
+</details>
+
+By default, the library object points to `nullptr`. In the current programming
+model in oneDNN, this object will point to a non-`nullptr` *only* in the
+following cases:
+1. within primitive execution whereby the user has passed a stream that contains
+   a Threadpool implementation, or
+2. the user has invoked `dnnl_sgemm_tp()` API by passing a pointer to a
+   Threadpool implementation.
+
+
+Similarly, in the current programming model in oneDNN, this object will point to
+`nullptr`
+1. Inside a parallel region,
+2. When the user invokes the `dnnl_sgemm()` API or `dnnl_sgemm_tp()` API
+   (without providing a Threadpool implementation) in the Threadpool build, and
+3. During primitive execution if the user passed a stream without a Threadpool
+   implementation.
+
+
+## Problem statement
+
+In the context of the Threadpool runtime, this RFC discusses two issues in the
+current programming model in oneDNN.
+
+### Primitive decomposition
+
+Because `active_threadpool` may or may not point to `nullptr` at a given point
+in time, certain *internal* functions in oneDNN will behave differently. For
+instance, `dnnl_get_max_threads()` will return
+`std::thread::hardware_concurrency()` if `active_threadpool == nullptr` when it
+is invoked. This function can be invoked at, e.g., primitive (descriptor)
+creation, primitive execution or in GEMM.
+
+In the programming model of an application that uses oneDNN, e.g. Tensorflow, a
+stream can be created (and, consequently, a Threadpool implementation is passed
+to the library) *after* primitive (descriptor) creation. This means that, at
+primitive (descriptor) creation, the library is unaware of a Threadpool
+implementation *yet*. Therefore, any decomposition logic in oneDNN primitives
+that relies on `dnnl_get_max_threads()` will be optimized for
+`std::thread::hardware_concurrency()` number of threads.
+
+However, during primitive execution the actual number of threads in the
+Threadpool may (and most likely will) be different than
+`std::thread::hardware_concurrency()`, meaning that the primitive may no longer
+be optimized for the actual number of threads
+`dnnl::threadpool_iface::get_num_threads()`. Testing indicates that this
+behavior may affect primitive execution during inference, i.e. with `mb = 1` on
+high throughput systems, such as AVX-512.
+
+As an experiment, limiting the return value of `dnnl_get_max_threads()` to the
+number of threads in the Threadpool yields noticeable performance improvements
+(in the order of 20-30%). Below is performance comparison data for ResNet50 and
+Googlenet_v1 convolutions on AVX-512, where either
+`std::thread()::hardware_concurrency()` or
+`dnnl::threadpool_iface::get_num_threads()` is used for primitive decomposition,
+and using `dnnl::threadpool_iface::get_num_threads()` at execution.
+
+![](decomposition_perf.png)
```

Comment:
Do we have a good understanding of why? Could it be that our threading heuristic are just using `get_max_thread` improperly? 
For example if we adapt our heuristic to not always use all available threads, but just the number of threads below `get_max_threads()` that maximizes the performance, could we be good?

Reply:
What do you mean by "using `get_max_thread improperly"?

For your second question, I don't think we can do anything. Since, at primitive (descriptor) creation, we *do not* know the actual number of threads in the threadpool used at execution, we have to rely on something. The current assumption is that the number of threads used in a threadpool passed at execution will *not* exceed `std::thread::hardware_concurrency()`. By doing this, we make sure that all the threads in the threadpool will have work to do during execution.

Using a number smaller than `std::thread::hardware_concurrency()` may lead to some threads doing no work

Reply:
Usually, using all the threads is not always better than using some of the threads.
For example, if I have to compute a matrix multiplication of two 2x2 matrices with 4 threads available, it will likely be faster to do it in one thread that to use 4 threads and have them synchronize to compute the reduction.
The same applies with larger sizes: more threads could imply smaller blocks, which has an impact on data reuse. It could also imply a higher overhead if parallel reduction happens.

What I meant by improperly is that maybe our current heuristics try to make all the threads work (as you mentioned). Historically, this was fine, since we were supporting only OpenMP and assumed that the number of threads was at most the number of cores. This assumption is not true anymore.

Reply:
I agree that too much decomposition is useless, that's what we do internally for GEMM.

However, the challenge of using an certain internal upper bound for number of threads will depend on the threading implementation. Since OpenMP is a standard, we can rely on it's abilities (such as busy waiting, synchronization) to improve performance.

The threadpool threading in oneDNN does not rely on a particular implementation. Therefore we can make only generic assumptions. For instance, our GEMM (and other primitives) will not use k-reduction for threadpool since the threads in the Eigen implementation are asynchronous. The only assumption I can think of here is not decomposing the problem if it fits entirely into L1 cache.

What other assumption you can think of we can use for our heuristics?

Reply:
Going back to your questions "Do we have a good understanding why?". I don't have a clear understanding why this performance issue is there, however I would assume it's due to inefficient memory usage. For OpenMP, I believe our decomposition strongly relies on high L1 usage for each block on each core, whereas having multiple blocks running on a single core may be a problem.

Reply:
k-reduction was just one example. With this one aside, I think the other big issue is data reuse: if you maximize threading, you use smaller blocks. This can decrease how much data gets reused from cache and:
1. cause more accesses to DDR.
2. or more sharing between cores, and communication between cores can become a bottleneck.

Even with general assumption on threading, implementations could try to avoid over decomposition by using information available from hardware registers (e.g. CPUID on x64 could provide cache sizes, number of physical cores, ...).

Reply:
Basically, my main concern is that this is potentially an implementation issue that we try to solve with an API.
Here by default, if the user does not pass that extra attribute, the performance will remain non-optimal.

Also, it puts the burden of getting best performance on the user. To get decent performance, the user will have to pass a proper value, and they might not know it (for example, if the threadpool is not yet created at primitive creation time). Or worse, users may try to "hack" heuristics to get best performance, but end up with worse performance over time as we modify our implementations.

Reply:
I would say that this is rather related to our programming model rather than an implementation issue. What we try to achieve here is to let the library know the number of threads available in the threadpool before execution. The library will still decide on the decomposition, but it will use a number of threads provided by the user as an upper limit instead of some generic assumption. We do not ask the user to tune this parameter to achieve a certain level of performance, or change the heuristics. I don't see how else we can change this without changing our programming model.

Reply:
I have introduced an additional Option (4), where I briefly present possibility of using internal heuristics to reduce thread usage for our primitive implementation. Let me know if you think this would work.